core/src/main/java/hudson/model/Label.java

```diff
@@ -611,9 +620,34 @@ public static Label get(String l) {
      *
      * TODO: replace this with a real parser later
      */
-    public static Label parseExpression(@NonNull String labelExpression) throws ANTLRException {
-        LabelExpressionLexer lexer = new LabelExpressionLexer(new StringReader(labelExpression));
-        return new LabelExpressionParser(lexer).expr();
+    public static Label parseExpression(@NonNull String labelExpression) throws ParseCancellationException {
```

Comment:
[Effective Java 3rd Edition §74](https://books.google.com/books?id=BIpDDwAAQBAJ) says:

> Use the Javadoc `@throws` tag to document each exception that a method can throw, but do _not_ use the `throws` keyword on unchecked exceptions.

Here and elsewhere, any instances of `throws ParseCancellationException` should be deleted in favor of a Javadoc `@throws` tag.

Also consider loosening the contract in the `@throws` tag to specify something further up in `ParseCancellationException`'s type hierarchy, such as `CancellationException` or `IllegalStateException` (and updating the `@deprecated` message in `ANTLRException` accordingly). The advantage in declaring the contract in terms of pure Java Platform types is that it prevents callers from relying on a type that is subject to change in the future. In other words, this would be a smart move that would make life easier for whoever is upgrading us from ANTLR4 to the next big thing. By defining our API in terms of pure Java types, we are free to throw a specific subtype in the implementation (like `org.antlr.v4.runtime.misc.ParseCancellationException`) but also leave ourselves the option of throwing some other type in the future (as long as it inherits from the Java Platform type we specify in the contract) rather than tying ourselves to the specific ANTLR4 implementation.